File: src/main/java/com/slavikjunior/orm/CRUD.java
package com.slavikjunior.orm;

import com.slavikjunior.annotations.Entity;
import kotlin.Pair;

import java.util.List;

public interface CRUD {

    <T extends CRUDable> void create(T entity, List<Object> properties);
    <T extends CRUDable> void read(T entity);
    <T extends CRUDable> void update(T entity);
    <T extends CRUDable> void delete(T entity);
}

File: src/main/java/com/slavikjunior/orm/CRUD.kt
package com.slavikjunior.orm

import com.slavikjunior.annotations.WrappedClass

interface CRUD {

    fun <T : CRUDable> createEntity(entity: T, idIsAutoGenerate: Boolean = true): Boolean?
    fun <T : CRUDable> getEntityById(entityClass: Class<T>, id: Int): T?
    fun <T : CRUDable, E> getEntityByPairsOfColumnsAndValues(entityClass: Class<T>, columnsToValues: Map<String, @WrappedClass E>): T?
    fun <T : CRUDable, E> update(entityClass: Class<T>, id: Int, columnsToValues: Map<String, @WrappedClass E>): Boolean
    fun <T : CRUDable, E> updateAndGet(entityClass: Class<T>, id: Int, columnsToValues: Map<String, @WrappedClass E>): T?
    fun <T : CRUDable> delete(entity: T?)
}